depthai/plugin.cpp

```diff
@@ -0,0 +1,344 @@
+            img0 = cam_type_.img0;
+            img1 = cam_type_.img1;
+            rgb = cam_type_.rgb;
+            depth = cam_type_.depth;
+        }
+        
+        std::chrono::time_point<std::chrono::steady_clock, std::chrono::steady_clock::duration> gyroTs;
+        Eigen::Vector3f la;
+        Eigen::Vector3f av;
+        
+        if(imuPacket_go) {
+            auto imuPacket = imuQueue->tryGet<dai::IMUData>();
+            #ifndef NDEBUG
+                if(imu_packet == 0){
+                    first_packet = std::chrono::steady_clock::now();
+                }
+                imu_packet++;
+            #endif
+            
+            auto imuDatas = imuPacket->imuDatas;
+            for(auto& imuData : imuDatas) {
+                if(useRaw){
+                    gyroTs = std::chrono::time_point_cast<std::chrono::nanoseconds>(imuData.rawGyroscope.timestamp.getTimestamp());
+                    if (gyroTs <= test_time_point){return;}
+                    test_time_point = gyroTs;
+                    la = {imuData.rawAcceleroMeter.x, imuData.rawAcceleroMeter.y, imuData.rawAcceleroMeter.z};
+                    av = {imuData.rawGyroscope.x, imuData.rawGyroscope.y, imuData.rawGyroscope.z};
+                }
+                else {
+                    gyroTs = std::chrono::time_point_cast<std::chrono::nanoseconds>(imuData.gyroscope.timestamp.getTimestamp());
+                    if (gyroTs <= test_time_point){return;}
+                    test_time_point = gyroTs;
+                    la = {imuData.acceleroMeter.x, imuData.acceleroMeter.y, imuData.acceleroMeter.z};
+                    av = {imuData.gyroscope.x, imuData.gyroscope.y, imuData.gyroscope.z};
+                }
+                
+            }
+        
+            // Time as ullong (nanoseconds)
+            ullong imu_time = static_cast<ullong>(gyroTs.time_since_epoch().count());
+
+            // Time as time_point
+            using time_point = std::chrono::system_clock::time_point;
+            time_type imu_time_point{std::chrono::duration_cast<time_point::duration>(std::chrono::nanoseconds(imu_time))};
+            
+            // Submit to switchboard
+            #ifndef NDEBUG
+                imu_pub++;
+            #endif
+            _m_imu_cam.put(_m_imu_cam.allocate<imu_cam_type>(
+                {
+                    imu_time_point,
+                    av,
+                    la,
+                    img0,
+                    img1,
+                    imu_time
+                }
+            ));
+            
+            if (rgb && depth)
+            {
+                #ifndef NDEBUG
+                    rgbd_pub++;
+                #endif
+                _m_rgb_depth.put(_m_rgb_depth.allocate<rgb_depth_type>(
+                    {
+                        rgb,
+                        depth,
+                        imu_time
+                    }
+                ));
+            }
+        }
```

Comment:
In the current implementation, we always have `rgb` and `depth` available (i.e. not `nullopt`) since we always copy them together with `img0` and `img1` (we wait for all for `_go`'s). If this is the desired behavior, then we can short-circuit this `callback` earlier if we haven't received all the data yet.

Reply:
I'm confused on what we are short circuiting and the gain. We always publish the ```imu_cam_type``` based on IMU data regardless of the images and if we don't have RGB/depth data we jump to the end of the callback.